crates/cxx-qt-build/src/diagnostics.rs

```diff
@@ -0,0 +1,142 @@
+// SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+//! This module takes care of displaying errors emitted by CXX-Qt as nicely-printed diagnostics
+//! using codespan-reporting.
+
+use std::{fmt::Display, ops::Range, path::PathBuf};
+
+use proc_macro2::LineColumn;
+
+/// We need to wrap the CXX and CXX-Qt errors in a single error type so that they
+/// can be returned from the generation phases. This then allows for us to not unwrap
+/// the CXX failures meaning that we can display the macro expansion correctly.
+#[derive(Debug)]
+pub(crate) enum GeneratedError {
+    Cxx(cxx_gen::Error),
+    CxxQt(cxx_qt_gen::Error),
+}
+
+impl Display for GeneratedError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            GeneratedError::Cxx(err) => write!(f, "{}", err),
+            GeneratedError::CxxQt(err) => write!(f, "{}", err),
+        }
+    }
+}
+
+impl From<cxx_gen::Error> for GeneratedError {
+    fn from(err: cxx_gen::Error) -> Self {
+        Self::Cxx(err)
+    }
+}
+
+impl From<cxx_qt_gen::Error> for GeneratedError {
+    fn from(err: cxx_qt_gen::Error) -> Self {
+        Self::CxxQt(err)
+    }
+}
+
+pub(crate) struct Diagnostic {
+    file_path: PathBuf,
+    error: GeneratedError,
+}
+
+impl Diagnostic {
+    pub(crate) fn new(file_path: PathBuf, error: GeneratedError) -> Self {
+        Self { file_path, error }
+    }
+
+    fn byte_span_in(&self, source: &str) -> Option<Range<usize>> {
+        match &self.error {
+            // TODO: CXX doesn't have span available yet
+            GeneratedError::Cxx(_) => None,
```

Comment:
We need to try and add `span()` to `cxx_gen::Error` upstream, as this will then allow us to point to the code

Reply:
This is #536